### PR TITLE
test parsnip "overhead"

### PR DESCRIPTION
**Pull request by simonpcouch**
Some minimal testing to alert us when we've allowed additional "overhead" to creep in. A more true test of overhead would test differences rather than ratios, but those numbers are largely system-dependent.

My intent here is not that we'd automatically reject a PR that causes this test to fail, but to let us know when a PR changes these ratios and give us a chance to consider whether the slowdown is worth the benefit or ought to be addressed.

### CODE REVIEW

File: tests/testthat/test_fit_interfaces.R [hunk 1]
```diff
@@ -154,3 +154,20 @@ test_that("fit() can handle attributes on a vector outcome", {
...
+  }
+
+  time_engine <-
+    timing(lm(mpg ~ ., mtcars))
+  time_parsnip_form <-
+    timing(fit(parsnip::linear_reg(), mpg ~ ., mtcars))
+  time_parsnip_xy <-
+    timing(fit_xy(parsnip::linear_reg(), mtcars[2:11], mtcars[1]))
+
+  expect_true(time_parsnip_form / time_engine < 4.5)
```
Comment (simonpcouch) [inline tests/testthat/test_fit_interfaces.R:171]: This ratio is around 3.8 on my laptop.

File: tests/testthat/test_fit_interfaces.R [hunk 2]
```diff
@@ -154,3 +154,20 @@ test_that("fit() can handle attributes on a vector outcome", {
...
+    expr <- substitute(expr)
+    system.time(replicate(100, eval(expr)))[["elapsed"]]
+  }
+
+  time_engine <-
+    timing(lm(mpg ~ ., mtcars))
+  time_parsnip_form <-
+    timing(fit(parsnip::linear_reg(), mpg ~ ., mtcars))
+  time_parsnip_xy <-
+    timing(fit_xy(parsnip::linear_reg(), mtcars[2:11], mtcars[1]))
```
Comment (simonpcouch) [inline tests/testthat/test_fit_interfaces.R:169]: The bracket subsetting does add some additional compute that's not "our fault," though it's a fraction of `time_engine`.

**Comment by simonpcouch** [comment]
ubuntu-latest (devel) failure is due to unrelated #1074.

**Comment by EmilHvitfeldt** [comment]
I feel a little weird trying to test these, but I agree about the sentiment. I totally agree about the skip on cran!

Firstly, how variable are the ratios when you test them locally?

**Comment by simonpcouch** [comment]
> Firstly, how variable are the ratios when you test them locally?

Heard. Here are the distributions of those ratios locally:

``` r
library(parsnip)

timing <- function(expr) {
  expr <- substitute(expr)
  system.time(replicate(100, eval(expr)))[["elapsed"]]
}

time_engine <-
  replicate(100, timing(lm(mpg ~ ., mtcars)))
time_parsnip_form <-
  replicate(100, timing(fit(linear_reg(), mpg ~ ., mtcars)))
time_parsnip_xy <-
  replicate(100, timing(fit_xy(linear_reg(), mtcars[2:11], mtcars[1])))

hist(time_parsnip_form / time_engine)
```



``` r
hist(time_parsnip_xy / time_engine)
```



<sup>Created on 2024-02-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by simonpcouch** [comment]
Given the current timings, we could even use 1000 replicates for each and this test would only take a couple seconds, if we wanted to cut down on that variability further. I've set the thresholds to be quite permissive generally, though.

**Comment by EmilHvitfeldt** [comment]
that is not a bad idea. Up to you 😄

File: tests/testthat/test_fit_interfaces.R [hunk 3]
```diff
@@ -154,3 +154,20 @@ test_that("fit() can handle attributes on a vector outcome", {
...
+
+  expect_true(time_parsnip_form / time_engine < 4.5)
+  expect_true(time_parsnip_xy / time_engine < 6)
```
[the remark this answers is not in the document]
Comment (topepo) [reply tests/testthat/test_fit_interfaces.R:175]: Only one was over the limit here: 

image

**Review by topepo — COMMENTED** [review]
Maybe instead of summing the elapsed times for 100 reps, take the median of the reps (or mean) and do a ratio of those.

File: tests/testthat/test_fit_interfaces.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonpcouch) [reply tests/testthat/test_fit_interfaces.R:175]: Is this a distribution of the timings of individual runs or of sums of 100 runs? Could you share the code that generated this?

**Comment by simonpcouch** [comment]
I like the idea of using the median timing! The only issue is that individual fits go very fast, and `system.time()` only measures in 1000ths of a second, so timings on individual runs would be either .0000 or .0001.

**Comment by topepo** [comment]
You gotta bump them numbers up `mtcars[rep(1:32, 100),]` or use a different data set (Sacramento?)

**Comment by simonpcouch** [comment]
The decreased ratio cutoffs in the above comment are due to speedups merged upstream from PRs in the last two days!

**Review by topepo — COMMENTED** [review]
Looks great with bench. Not much dependency overhead either.

**Pull request merged**